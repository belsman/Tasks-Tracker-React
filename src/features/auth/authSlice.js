import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  authToken: null,
  status: 'idle',
};

export const registerUserAsync = createAsyncThunk(
  'auth/registerUser',
  async signupData => {
    const response = await axios({
      method: 'post',
      url: 'http://127.0.0.1:9000/signup',
      data: signupData,
      headers: { 'Content-Type': 'application/json' },
    });
    const data = await response.data;
    return data.auth_token;
  },
);

export const loginUserAsync = createAsyncThunk(
  'auth/loginUser',
  async loginData => {
    const response = await axios({
      method: 'post',
      url: 'http://127.0.0.1:9000/auth/login',
      data: loginData,
      headers: { 'Content-Type': 'application/json' },
    });
    const data = await response.data;
    return data.auth_token;
  },
);

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  /* eslint-disable no-param-reassign */
  reducers: {
    logout: state => {
      state.authToken = null;
      state.status = 'idle';
    },
  },

  extraReducers: builder => {
    builder
      .addCase(registerUserAsync.pending, state => {
        state.status = 'loading';
      })
      .addCase(registerUserAsync.fulfilled, (state, action) => {
        state.status = 'idle';
        state.authToken = action.payload;
      })
      .addCase(loginUserAsync.pending, state => {
        state.status = 'loading';
      })
      .addCase(loginUserAsync.fulfilled, (state, action) => {
        state.status = 'idle';
        state.authToken = action.payload;
      });
  }, /* eslint-enable no-param-reassign */
});

export const { logout } = authSlice.actions;

export const isUserLogged = state => Boolean(state.auth.authToken);

export default authSlice.reducer;
