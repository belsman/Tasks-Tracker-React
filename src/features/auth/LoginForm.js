import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { loginUserAsync } from './authSlice';
import style from './auth.module.css';

const LoginForm = () => {
  const dispatch = useDispatch();
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');

  const submitHandler = e => {
    e.preventDefault();
    const bodyFormData = new FormData();
    bodyFormData.append('username', username);
    bodyFormData.append('password', password);
    dispatch(loginUserAsync(bodyFormData));
  };

  return (
    <form className={style.formELement} name="registration-form" method="POST" onSubmit={submitHandler}>
      <label className={style.formLabel} htmlFor="username">
        Username
        <input
          className={style.formGroup}
          type="text"
          id="username"
          value={username}
          onChange={e => setUsername(e.target.value)}
          placeholder="username"
        />
      </label>
      <label className={style.formLabel} htmlFor="password">
        Password
        <input
          className={style.formGroup}
          type="password"
          id="password"
          value={password}
          onChange={e => setPassword(e.target.value)}
          placeholder="password"
        />
      </label>
      <button className={style.btn} type="submit">Login</button>
    </form>
  );
};

export default LoginForm;
