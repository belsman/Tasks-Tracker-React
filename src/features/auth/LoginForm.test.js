import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import pretty from 'pretty';
import { act } from 'react-dom/test-utils';

/* eslint-disable no-unused-vars */
import { useDispatch } from 'react-redux';
import style from './auth.module.css';
import { loginUserAsync } from './authSlice';
/* eslint-enable no-unused-vars */

import LoginForm from './LoginForm';

jest.mock('./authSlice', () => ({
  loginUserAsync(fn) {
    return fn;
  },
}));

jest.mock('react-redux', () => ({
  useDispatch(fn) {
    return fn;
  },
}));

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
  act(() => {
    render(<LoginForm />, container);
  });
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('renders a Login input username field', () => {
  const username = document.querySelector('input#username');
  expect(username).not.toBeNull();
});

it('renders a Login password field', () => {
  const password = document.querySelector('input#password');
  expect(password).not.toBeNull();
});

it('should render the expected output', () => {
  act(() => {
    render(<LoginForm />, container);
  });

  expect(
    pretty(container.innerHTML),
  ).toMatchSnapshot();
});
