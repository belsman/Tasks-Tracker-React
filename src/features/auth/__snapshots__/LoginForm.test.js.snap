// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the expected output 1`] = `"<form class=\\"formELement\\" name=\\"registration-form\\" method=\\"POST\\"><label class=\\"formLabel\\" for=\\"username\\">Username<input class=\\"formGroup\\" type=\\"text\\" id=\\"username\\" placeholder=\\"username\\" value=\\"\\"></label><label class=\\"formLabel\\" for=\\"password\\">Password<input class=\\"formGroup\\" type=\\"password\\" id=\\"password\\" placeholder=\\"password\\" value=\\"\\"></label><button class=\\"btn\\" type=\\"submit\\">Login</button></form>"`;
